### ISSUE TITLE
Add way of composing an objective including a loss and regularization penalty

### ISSUE THREAD

**Issue by zackchase** [sharadmv/deepx#11]
Right now mixins combine an architecture with a predict function, loss, and optimizer.

This approach has a couple problems

1. The predict function shouldn't need to be tied to a particular loss or optimizer. 

2. What if I ant to train a model with square loss and l2 regularization. There is no way at present to combine loss functions with penalties within the shorthand.
..*  One way is to pass in a list of objectives [cross_entropy, weight_decay(1e-4)]. The optimization objective would then be the sum of these terms. 

3. Regarding prediction from optimizer etc, I think we should consider renaming arch and model to model and trainer. Model would have a predict() fn. model.predict() would not compile automatically but instead would compile on the first call to the function.
